Sources/Document/Json/JSONTree.swift

```diff
@@ -0,0 +1,505 @@
+        if lhs.type != rhs.type {
+            return false
+        }
+
+        return true
+    }
+
+    init(type: TreeNodeType, attributes: [String: String]? = nil, children: [any JSONTreeNode] = []) {
+        self.type = type
+        self.attributes = attributes
+        self.children = children
+    }
+}
+
+/**
+ * `TextNode` is a node that has a value.
+ */
+public struct TextNode: JSONTreeNode {
+    public let type = DefaultTreeNodeType.text.rawValue
+    public let value: String
+}
+
+/**
+ * `buildDescendants` builds descendants of the given tree node.
+ */
+func buildDescendants(treeNode: any JSONTreeNode, parent: CRDTTreeNode, context: ChangeContext) throws {
+    let ticket = context.issueTimeTicket
+
+    if let node = treeNode as? TextNode {
+        let textNode = CRDTTreeNode(pos: CRDTTreePos(createdAt: ticket, offset: 0), type: DefaultTreeNodeType.text.rawValue, value: node.value)
+
+        try parent.append(contentsOf: [textNode])
+    } else if let node = treeNode as? ElementNode {
+        let attrs = RHT()
+
+        node.attributes?.forEach { key, value in
+            attrs.set(key: key, value: value, executedAt: ticket)
+        }
+
+        let elementNode = CRDTTreeNode(pos: CRDTTreePos(createdAt: ticket, offset: 0), type: node.type, attributes: attrs.size == 0 ? nil : attrs)
+
+        try parent.append(contentsOf: [elementNode])
+
+        try node.children.forEach { child in
+            try buildDescendants(treeNode: child, parent: elementNode, context: context)
+        }
+    } else {
+        throw YorkieError.unexpected(message: "Must not here!")
+    }
+}
+
+/**
+ * createCRDTTreeNode returns CRDTTreeNode by given TreeNode.
+ */
+func createCRDTTreeNode(context: ChangeContext, content: any JSONTreeNode) throws -> CRDTTreeNode {
+    let ticket = context.issueTimeTicket
+
+    let root: CRDTTreeNode
+    if let node = content as? TextNode {
+        root = CRDTTreeNode(pos: CRDTTreePos(createdAt: ticket, offset: 0), type: node.type, value: node.value)
+    } else if let node = content as? ElementNode {
+        let attrs = RHT()
+
+        node.attributes?.forEach { key, value in
+            attrs.set(key: key, value: value, executedAt: ticket)
+        }
+
+        root = CRDTTreeNode(pos: CRDTTreePos(createdAt: ticket, offset: 0), type: node.type, attributes: attrs.size == 0 ? nil : attrs)
+
+        try node.children.forEach { child in
+            try buildDescendants(treeNode: child, parent: root, context: context)
+        }
+    } else {
+        throw YorkieError.unexpected(message: "Must not here!")
+    }
+
+    return root
+}
+
+/**
+ * `JSONTree` is a CRDT-based tree structure that is used to represent the document
+ * tree of text-based editor such as ProseMirror.
+ */
+public class JSONTree {
+    private let initialRoot: ElementNode?
+    private var context: ChangeContext?
+    private var tree: CRDTTree?
+
+    public convenience init() {
+        self.init(initialRoot: nil)
```

Comment:
I think it is okay to remove the convenience initializer since `public init(initialRoot:)` already has a default parameter value of nil.

Reply:
Right. Thanks.